### PR TITLE
refactor(sdk): introduce CredentialsProvider to manage and apply credentials for http requests

### PR DESCRIPTION
**Pull request by elainaRenee**
Description
-----------

- Fixes WB-NNNNN
- Fixes #NNNN

What does the PR do? Include a concise description of the PR contents.

- [ ] I updated CHANGELOG.md, or it's not applicable

Testing
-------
I installed the wandb-core in development mode by running `./core/scripts/setup-core-path.sh` and tested that I am able to authenticate with an api key

### CODE REVIEW

File: core/pkg/server/connection.go [hunk 1]
```diff
@@ -268,24 +268,6 @@ func (nc *Connection) handleServerRequest() {
...
-
-	err := settings.EnsureAPIKey()
-	if err != nil {
-		slog.Error(
-			"connection: couldn't get API key",
-			"err", err,
-			"id", nc.id,
-		)
-		panic(err)
-	}
```
Comment (elainaRenee) [inline core/pkg/server/connection.go:287]: I moved this to the credentials provider, which gets created when a backend is created

File: core/pkg/server/stream_init.go [hunk 2]
```diff
@@ -39,10 +39,17 @@ func NewBackend(
 		logger.CaptureFatalAndPanic(
 			fmt.Errorf("sender: failed to parse base URL: %v", err))
 	}
+
+	credentialProvider, err := api.NewCredentialProvider(settings)
+	if err != nil {
+		logger.CaptureFatalAndPanic(
+			fmt.Errorf("sender: failed to fetch credentials: %v", err))
```
Comment (kptkin) [inline core/pkg/server/stream_init.go:46]:
```suggestion
			fmt.Errorf("stream_init: failed to fetch credentials: %v", err))
```

File: core/internal/api/credentials.go [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+	"encoding/base64"
+	"fmt"
+	"log/slog"
+	"net/http"
+
+	"github.com/wandb/wandb/core/internal/settings"
+)
+
+// CredentialProvider defines an interface for managing and applying credentials
+// on HTTP requests
```
Comment (timoffex) [inline core/internal/api/credentials.go:13]: how about

```suggestion
// CredentialProvider adds credentials to HTTP requests.
```

File: core/internal/api/credentials.go [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+			"upcoming release. To continue using this feature, please " +
+			"downgrade to version 0.17.9 or lower using the following " +
+			"command: pip install wandb==0.17.9. Thank you for your patience.")
+	}
+	return NewAPIKeyCredentialProvider(settings)
+}
+
+var _ CredentialProvider = &APIKeyCredentialProvider{}
+
+type APIKeyCredentialProvider struct {
```
Comment (timoffex) [inline core/internal/api/credentials.go:35]: nit: could leave the concrete struct private

```go
type apiKeyCredentialProvider
```

File: core/internal/api/credentials.go [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+	apiKey string
+}
+
+func NewAPIKeyCredentialProvider(
+	settings *settings.Settings) (CredentialProvider, error) {
```
Comment (timoffex) [inline core/internal/api/credentials.go:40]:
```suggestion
func NewAPIKeyCredentialProvider(
	settings *settings.Settings,
) (CredentialProvider, error) {
```

**Review by timoffex — APPROVED** [review]

File: core/internal/api/credentials.go [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+
+func NewAPIKeyCredentialProvider(
+	settings *settings.Settings,
+) (CredentialProvider, error) {
+	if err := settings.EnsureAPIKey(); err != nil {
+		slog.Error(
+			"connection: couldn't get API key",
+			"err", err,
+		)
+		return nil, err
```
Comment (kptkin) [inline core/internal/api/credentials.go:47]: could you just return the error and log it using our logger in the stream_init call-site?

Comment (elainaRenee) [reply core/internal/api/credentials.go]: @kptkin done

**Review by kptkin — APPROVED** [review]
So much better! Thanks!

**Pull request merged**